File: RoadRunnerJS/src/jsMain/kotlin/trajectory/constraints/AngularVelocityConstraint.kt
package trajectory.constraints

import geometry.Pose2d
import kotlin.js.JsExport
import kotlin.math.abs
import kotlin.math.max

/**
 * Constraint limiting angular velocity.
 */
@JsExport
class AngularVelocityConstraint(
    private val maxAngularVel: Double
) : TrajectoryVelocityConstraint() {
    override fun get(s: Double, pose: Pose2d, deriv: Pose2d, baseRobotVel: Pose2d): Double {
        val omega0 = baseRobotVel.heading
        if (abs(omega0) >= maxAngularVel) {
            throw UnsatisfiableConstraint()
        }

        return max(
            (maxAngularVel - omega0) / deriv.heading,
            (-maxAngularVel - omega0) / deriv.heading
        )
    }
}

File: src/roadrunnerjs/trajectory/constraints/AngularVelocityConstraint.ts
import { Pose2d } from "roadrunnerjs/geometry";
import { TrajectoryVelocityConstraint } from "./TrajectoryVelocityConstraint";
/**
 * Constraint limiting angular velocity.
 */
export class AngularVelocityConstraint implements TrajectoryVelocityConstraint {
  constructor(private readonly maxAngularVel: number) {}

  get(s: number, pose: Pose2d, deriv: Pose2d, baseRobotVel: Pose2d): number {
    const omega0 = baseRobotVel.heading;
    if (Math.abs(omega0) >= this.maxAngularVel) {
      throw new Error("Unsatisfiable Constraint");
    }

    return Math.max(
      (this.maxAngularVel - omega0) / deriv.heading,
      (-this.maxAngularVel - omega0) / deriv.heading
    );
  }
}
